server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -47,234 +40,74 @@
-                    logger.debug("[{}] no longer dangling (created), removing from dangling list", index);
-                }
-                danglingIndices.remove(index);
-            }
-        }
+        final Metadata metadata = this.clusterService.state().metadata();
+        return findDanglingIndices(metadata);
     }
 
-    /**
-     * Finds (@{link #findNewAndAddDanglingIndices}) and adds the new dangling indices
-     * to the currently tracked dangling indices.
-     */
-    void findNewAndAddDanglingIndices(final Metadata metadata) {
-        final IndexGraveyard graveyard = metadata.indexGraveyard();
+    // Extracted from getDanglingIndices() as a package-private method to allow easier testing testing
+    Map<Index, IndexMetadata> findDanglingIndices(Metadata metadata) {
+        final Set<String> excludeIndexPathIds = new HashSet<>(metadata.indices().size());
 
-        // If a tombstone is created for a dangling index, we need to make sure that the
-        // index is no longer considered dangling.
-        danglingIndices.keySet().removeIf(graveyard::containsIndex);
-
-        danglingIndices.putAll(findNewDanglingIndices(danglingIndices, metadata));
-    }
-
-    /**
-     * Finds new dangling indices by iterating over the indices and trying to find indices
-     * that have state on disk, but are not part of the provided metadata, or not detected
-     * as dangled already.
-     */
-    public Map<Index, IndexMetadata> findNewDanglingIndices(Map<Index, IndexMetadata> existingDanglingIndices, final Metadata metadata) {
-        final Set<String> excludeIndexPathIds = new HashSet<>(metadata.indices().size() + danglingIndices.size());
         for (ObjectCursor<IndexMetadata> cursor : metadata.indices().values()) {
             excludeIndexPathIds.add(cursor.value.getIndex().getUUID());
         }
-        for (Index index : existingDanglingIndices.keySet()) {
-            excludeIndexPathIds.add(index.getUUID());
-        }
+
         try {
             final List<IndexMetadata> indexMetadataList = metaStateService.loadIndicesStates(excludeIndexPathIds::contains);
-            Map<Index, IndexMetadata> newIndices = new HashMap<>(indexMetadataList.size());
+            final Map<Index, IndexMetadata> danglingIndices = new HashMap<>();
             final IndexGraveyard graveyard = metadata.indexGraveyard();
 
             for (IndexMetadata indexMetadata : indexMetadataList) {
                 Index index = indexMetadata.getIndex();
                 if (graveyard.containsIndex(index) == false) {
-                    newIndices.put(index, stripAliases(indexMetadata));
+                    danglingIndices.put(index, stripAliases(indexMetadata));
                 }
             }
 
-            return newIndices;
+            return danglingIndices;
         } catch (IOException e) {
-            logger.warn("failed to list dangling indices", e);
+            logger.warn("Failed to list dangling indices", e);
```

Comment:
I prefer the original:

```suggestion
            logger.warn("failed to list dangling indices", e);
```